README
% README file for tin-pot/discount   
% Martin Hofmann <tin-pot@gmx.net>  
% 2014-02-12  


The Markdown tool "discount" on Windows: `tin-pot/discount`
===========================================================

This repository is a clone of [David Parsons' implementation][dp] in C
of (a processor for) [John Gruber's "Markdown" markup language][md],
with added support to build the package with Visual Studio on Windows.
There are also some added command line options of the tools and one new
flag value of the library.

See also the home page of this distribution at <https://github.com/tin-
pot/discount>.

In the following, the term "*Unix*" refers to all the Unix-like
platforms that `discount` does target originally.


1  Modifications
----------------

The modifications added in this distribution are:

- Added pre-configured files

  - `version.c`,
  - `mkdio.h`, and
  - `config.h`,

  in `MSVC/` so that no `configure.sh` run---which generates these files
  on Unix---is needed on Windows.

- Added project files for all executables and libraries in `MSVC/`.

- Added supporting files for building on Windows in `MSVC/posc/`,
  including the [`freegetopt`][go] implementation of Unix's `getopt()`
  function which is used by `main.c`.

See the [CHANGES](CHANGES.html) file for more details.


2  Building/Installing
----------------------

The package should build out of the box with Visual Studio 2003.NET or
Visual Studio 2008 or newer (later versions of Visual Studio will import
and convert the project files).

The generated executable are statically linked: they only depend on
`kernel32.dll`. Place them wherever you like on your `PATH`.


3  About the Port
-----------------

Porting to Windows was done with a fairly general technique, which
presents to the compiler a "faked" environment of header files on
top of the "real" header files of MSVC. All this is contained in the
`MSVC/posc` directory. See the `MSVC/posc/README.html` file there for
explanations.

All `discount` source files in the project root directory are
unmodified. Their need for functions like `strncasecmp()` (Unix), or
`vsnprintf()` (C99), or `getopt()` gets satisfied completely out of
files in `posc`.

The included `getopt()` implementation is [freegetopt][go], its license
is in the file `LICENSE-freegetopt`.


[dp]:http://www.pell.portland.or.us/~orc/Code/discount/
[md]:http://daringfireball.net/projects/markdown/
[go]:http://freegetopt.sourceforge.net/



CHANGES
% CHANGES file of tin-pot/discount   
% Martin Hofmann <tin-pot@gmx.net>  
% 2014-02-12  


Changes in this distribution of `discount`
==========================================

1  Overview
-----------

For the most part, the `discount` sources are identical to these in
David Parsons' `discount-2.1.6.tar.bz2` source archive downloadable from
his [project site][dp].

Building upon this, this distribution adds the following features:

1. Project files for MS Visual C++ 2003 and 2008 (in `MSVC/`, as well as

2. files to allow building the Unixish `discount` package on Windows (in
   `MSVC/posc/` without additional requirements.

3. The new options `-S` and `-I` to `mkd2html.c`, to choose *HTML 4.01
   Strict* and *ISO 15445:2000* doctypes,

4. and a fix to a [bug][co] in `mkd2html.c`.

5. The new flag value `wiki` (with the value `0x00800000`) to produce a
   somewhat Wiki-friendly variant of output HTML.


2  Changes to `discount` source files
-------------------------------------

`generate.c`

:   **ADD:** Added code tests `MKD_WIKI` and `IS_WIKI`; new function
    `wikiurl()`; new *linkytype* `wikit` and related `IS_WIKI` macro.

`markdown.h`

:   **ADD:** Add definition of `MKD_WIKI` (**Important:** must match
    definition in `mkdio.h.in`!)

`mkd2html.c`

:   **FIX:** `while ( argc > 2 )` (see [*github* commit][co].)

`pgm_options.c`

:   **ADD:** Detect flag `-f wiki`.


3  Changes to `discount` template files
---------------------------------------

The configuration script `configure.sh` generates three files:

1. `config.h` from user options and platform diagnostics.
2. `Makefile` from the template `Makefile.in`.
3. `version.c` from the template `version.c.in`.
4. `mkdio.h` from the template `mkdio.h.in`>

Required changes to these generated files are effected by changing the
corresponding template file:

`version.c.in`

:    **ADD:** A suffix added to the version identification shown in the
     `markdown -V` output: `"-wiki"`. This differentiates it from the
     version information shown by an *original* `markdown` built from
     `discount-2.1.6.tar.bz2`.

`mkdio.h.in`

:    **ADD:** Define the flag bit-value `MKD_WIKI`.


4  Files of `discount` explicitly *not* changed
-----------------------------------------------

`VERSION`

:    This should only contain a numeric version identification.
     The `markdown -V` version information output is modified in
     `version.c.in`.


5  Added files and directories
------------------------------

`MSVC/`

:    All files in this directory are new, written by Martin Hofmann
     (<tin-pot@gmx.net>), except these files in `MSVC/posc/`:

     - `ansidecl.h`, `ansidecl_1996.h`, and `ansidecl_2001.h` are
       *Copyright (c) 1991--2001 Free Software Foundation*.
     - `getopt.c` and `getopt.h` are *Copyright (c) 2000--2003 Mark K.
       Kim*.
     - `inttypes.h`, `stdint.h` are *Copyright (c) 2006 Alexander
       Chemeris*.

     These files contain their own license conditions, see the contents
     of `MSVC/posc/`.


[dp]:http://www.pell.portland.or.us/~orc/Code/discount/
[co]:https://github.com/Orc/discount/commit/c1904052041dafc085791e6f06340dbd1472536b



mkd2html.c
/*
 * mkd2html:  parse a markdown input file and generate a web page.
 *
 * usage:  mkd2html [options] filename
 *  or     mkd2html [options] < markdown > html
 *
 *  options
 *         -css css-file
 *         -header line-to-add-to-<HEADER>
 *         -footer line-to-add-before-</BODY>
 *
 * example:
 *
 *   mkd2html -cs /~orc/pages.css syntax
 *     ( read syntax OR syntax.text, write syntax.html )
 */
/*
 * Copyright (C) 2007 David L Parsons.
 * The redistribution terms are provided in the COPYRIGHT file that must
 * be distributed with this source code.
 */
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef HAVE_BASENAME
# ifdef HAVE_LIBGEN_H
#  include <libgen.h>
# else
#  include <unistd.h>
# endif
#endif
#include <stdarg.h>

#include "mkdio.h"
#include "cstring.h"
#include "amalloc.h"

char *pgm = "mkd2html";

#ifndef HAVE_BASENAME
char *
basename(char *path)
{
    char *p;

    if ( p = strrchr(path, '/') )
	return 1+p;
    return path;
}
#endif

void
fail(char *why, ...)
{
    va_list ptr;

    va_start(ptr,why);
    fprintf(stderr, "%s: ", pgm);
    vfprintf(stderr, why, ptr);
    fputc('\n', stderr);
    va_end(ptr);
    exit(1);
}


main(argc, argv)
char **argv;
{
    char *h;
    char *source = 0, *dest = 0;
    MMIOT *mmiot;
    int i;
    FILE *input, *output; 
    STRING(char*) css, headers, footers;
    mkd_flag_t outenc = MKD_OUT_UTF8;
    mkd_flag_t inenc  = MKD_IN_UTF8;
    mkd_flag_t flags = 0;
    const char *charset;
    enum { Transitional, Strict, Iso } doctype = Transitional;
    const char *const docdecl[] = {
	"-//W3C//DTD HTML 4.01 Transitional//EN", 
	"-//W3C//DTD HTML 4.01//EN", 
	"ISO/IEC 15445:2000//DTD HTML//EN"
    };
    const char *const docdtd[] = {
	" \"http://www.w3.org/TR/html4/loose.dtd\"",
	" \"http://www.w3.org/TR/html4/strict.dtd\"",
	""
    };


    CREATE(css);
    CREATE(headers);
    CREATE(footers);
    pgm = basename(argv[0]);

    while ( argc > 1) {
	if ( argc > 2 && strcmp(argv[1], "-css") == 0 ) {
	    EXPAND(css) = argv[2];
	    argc -= 2;
	    argv += 2;
	}
	else if ( argc > 2 && strcmp(argv[1], "-header") == 0 ) {
	    EXPAND(headers) = argv[2];
	    argc -= 2;
	    argv += 2;
	}
	else if ( argc > 2 && strcmp(argv[1], "-footer") == 0 ) {
	    EXPAND(footers) = argv[2];
	    argc -= 2;
	    argv += 2;
	}
	else if ( argv[1][0] == '-' ) {
	    char ch, *opt = argv[1];
	    
	    while ((ch = *++opt) != '\0') switch (ch) {
	    case 'S': doctype = Strict; continue;
	    case 'I': doctype = Iso;    continue;
	    case 'A': outenc = MKD_OUT_ASCII; continue;
	    case 'L': outenc = MKD_OUT_LATIN1; continue;
	    case 'U': outenc = MKD_OUT_UTF8; continue;
	    case 'l': inenc  = MKD_IN_LATIN1; continue;
	    default: 
		fprintf(stderr, 
		"usage: %s [-S | -I] [-A | -L | -U ] [-l] "
		"[-css URL] [-header text] [-footer text] [ source [dest] ]\n",
		 pgm);
		exit(1);
	    }
	    argc -= 1;
	    argv += 1;
	}
	else
	    break;
    }

    switch ( argc ) {
	char *p, *dot;
    case 1:
	input = stdin;
	output = stdout;
	break;
    case 2:
    case 3:
	dest   = malloc(strlen(argv[argc-1]) + 6);
	source = malloc(strlen(argv[1]) + 6);

	if ( !(source && dest) )
	    fail("out of memory allocating name buffers");

	strcpy(source, argv[1]);
	if (( p = strrchr(source, '/') ))
	    p = source;
	else
	    ++p;

	if ( (input = fopen(source, "r")) == 0 ) {
	    strcat(source, ".text");
	    if ( (input = fopen(source, "r")) == 0 )
		fail("can't open either %s or %s", argv[1], source);
	}
	strcpy(dest, source);

	if (( dot = strrchr(dest, '.') ))
	    *dot = 0;
	strcat(dest, ".html");

	if ( (output = fopen(dest, "w")) == 0 )
	    fail("can't write to %s", dest);
	break;

    default:
	fprintf(stderr,
		"usage: %s [-S | -I] [-A | -L | -U ] [-l] "
		"[-css URL] [-header text] [-footer text] [ source [dest] ]\n",
		 pgm);
	exit(1);
    }

    switch (outenc) {
    case MKD_OUT_ASCII:  charset = "US-ASCII";
			 flags |= MKD_OUT_ASCII;
			 break;
    case MKD_OUT_LATIN1: charset = "ISO-8859-1";
			 flags |= MKD_OUT_LATIN1;
			 break;
    case MKD_OUT_UTF8: default:
			 charset = "UTF-8";
			 flags |= MKD_OUT_UTF8;
			 break;
    }
    
    switch (inenc) {
    case MKD_IN_LATIN1:  flags |= MKD_IN_LATIN1;
                         break;
    case MKD_IN_UTF8: default:
                         flags |= MKD_IN_UTF8;
                         break;
    }
    
    if ( (mmiot = mkd_in(input, flags)) == 0 )
	fail("can't read %s", source ? source : "stdin");

    if ( !mkd_compile(mmiot, flags) )
	fail("couldn't compile input");


    h = mkd_doc_title(mmiot);

    /* print a header */

    fprintf(output,
	"<!doctype html public \"%s\"%s>\n"
	"<html>\n"
	"<head>\n"
	"  <meta name=\"GENERATOR\" content=\"mkd2html %s\">\n",
	docdecl[doctype],
	docdtd[doctype],
	markdown_version);
    
    fprintf(output,"  <meta http-equiv=\"Content-Type\"\n"
		   "        content=\"text/html; charset=%s\">\n",
		   charset);

    for ( i=0; i < S(css); i++ )
	fprintf(output, "  <link rel=\"stylesheet\"\n"
			"        type=\"text/css\"\n"
			"        href=\"%s\">\n", T(css)[i]);

    if ( h ) {
	fprintf(output,"  <title>");
	mkd_generateline(h, strlen(h), output, flags);
	fprintf(output, "</title>\n");
    }
    for ( i=0; i < S(headers); i++ )
	fprintf(output, "  %s\n", T(headers)[i]);
    fprintf(output, "</head>\n"
		    "<body>\n");

    /* print the compiled body */

    mkd_generatehtml(mmiot, output);

    for ( i=0; i < S(footers); i++ )
	fprintf(output, "%s\n", T(footers)[i]);
    
    fprintf(output, "</body>\n"
		    "</html>\n");
    
    mkd_cleanup(mmiot);
    exit(0);
}
